File: python/vstolib/variant.py
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


"""
The purpose of this python3 script is to implement the Variant dataclass.
"""


import statistics
import pandas as pd
from collections import defaultdict, OrderedDict
from dataclasses import dataclass, field
from typing import List, Dict, Tuple, ClassVar
from bisect import bisect_left, bisect_right, insort
from .variant_call_annotation import VariantCallAnnotation
from .variant_call import VariantCall


@dataclass
class Variant:
    id: str
    variant_calls: List[VariantCall] = field(default_factory=list)

    @property
    def alternate_alleles(self) -> List[str]:
        return [i.alternate_allele for i in self.variant_calls]

    @property
    def alternate_allele_read_counts(self) -> List[int]:
        return [i.alternate_allele_read_count for i in self.variant_calls]

    @property
    def alternate_allele_fractions(self) -> List[float]:
        return [i.alternate_allele_fraction for i in self.variant_calls]

    @property
    def alternate_allele_read_ids(self) -> List[List[str]]:
        return [list(i.alternate_allele_read_ids) for i in self.variant_calls]

    @property
    def attributes(self) -> List[OrderedDict]:
        return [i.attributes for i in self.variant_calls]

    @property
    def chromosome_1(self) -> str:
        return self.variant_calls[0].chromosome_1

    @property
    def chromosome_2(self) -> str:
        return self.variant_calls[0].chromosome_2

    @property
    def filter(self) -> List[str]:
        return [i.filter for i in self.variant_calls]

    @property
    def nucleic_acids(self) -> List[str]:
        return [i.nucleic_acid for i in self.variant_calls]

    @property
    def num_variant_calls(self) -> int:
        return len(self.variant_calls)

    @property
    def position_1(self) -> List[int]:
        return [i.position_1 for i in self.variant_calls]

    @property
    def position_1_annotations(self) -> List[List[VariantCallAnnotation]]:
        return [i.position_1_annotations for i in self.variant_calls]

    @property
    def position_1_stdev(self) -> float:
        return 0.0 if len(self.position_1) == 1 else statistics.stdev(self.position_1)

    @property
    def position_2(self) -> List[int]:
        return [i.position_2 for i in self.variant_calls]

    @property
    def position_2_annotations(self) -> List[List[VariantCallAnnotation]]:
        return [i.position_2_annotations for i in self.variant_calls]

    @property
    def position_2_stdev(self) -> float:
        return 0.0 if len(self.position_2) == 1 else statistics.stdev(self.position_2)

    @property
    def precise(self) -> List[bool]:
        return [i.precise for i in self.variant_calls]

    @property
    def quality_scores(self) -> List[float]:
        return [i.quality_score for i in self.variant_calls]

    @property
    def reference_alleles(self) -> List[str]:
        return [i.reference_allele for i in self.variant_calls]

    @property
    def reference_allele_read_counts(self) -> List[int]:
        return [i.reference_allele_read_count for i in self.variant_calls]

    @property
    def sample_ids(self) -> List[str]:
        return [i.sample_id for i in self.variant_calls]

    @property
    def sequencing_platform(self) -> List[str]:
        return [i.sequencing_platform for i in self.variant_calls]

    @property
    def source_ids(self) -> List[str]:
        return [i.source_id for i in self.variant_calls]

    @property
    def tags(self) -> List[List[str]]:
        return [list(i.tags) for i in self.variant_calls]

    @property
    def total_read_counts(self) -> List[int]:
        return [i.total_read_count for i in self.variant_calls]

    @property
    def variant_call_ids(self) -> List[str]:
        return [variant_call.id for variant_call in self.variant_calls]

    @property
    def variant_calling_methods(self) -> List[str]:
        return [i.variant_calling_method for i in self.variant_calls]

    @property
    def variant_sequences(self) -> List[List[str]]:
        return [list(i.variant_sequences) for i in self.variant_calls]

    @property
    def variant_sizes(self) -> List[int]:
        return [i.variant_size for i in self.variant_calls]

    @property
    def variant_size_stdev(self) -> float:
        return 0.0 if len(self.variant_sizes) == 1 else statistics.stdev(self.variant_sizes)

    @property
    def variant_subtypes(self) -> List[str]:
        return [i.variant_subtype for i in self.variant_calls]

    @property
    def variant_types(self) -> List[str]:
        return [i.variant_type for i in self.variant_calls]

    def __eq__(self, other):
        if isinstance(other, Variant):
            return (self.id) == (other.id)
        return NotImplemented

    def add_variant_call(self, variant_call: VariantCall):
        """
        Add a VariantCall object.
        """
        if len(self.variant_calls) > 0:
            if variant_call.chromosome_1 != self.variant_calls[0].chromosome_1 or \
                    variant_call.chromosome_2 != self.variant_calls[0].chromosome_2:
                raise Exception('VariantCall chromosome_1 and chromosome_2 must be the same as '
                                'existing VariantCall objects in this Variant object.')
        insort(self.variant_calls, variant_call)

    def to_dataframe_row(self) -> Dict:
        data = defaultdict(list)
        for variant_call in self.variant_calls:
            data['variant_id'].append(self.id)
            for key, value in variant_call.to_dataframe_row().items():
                data[key].append(value[0])
        return data

    def to_dataframe(self) -> pd.DataFrame:
        return pd.DataFrame(self.to_dataframe_row())

    def to_dict(self) -> Dict:
        data = {
            'id': self.id,
            'variant_calls': [variant_call.to_dict() for variant_call in self.variant_calls]
        }
        return data

File: src/variant.rs
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.


extern crate exitcode;
extern crate serde;
use serde::{Serialize, Deserialize};
use std::process;
use crate::variant_call::VariantCall;
use crate::utilities;


#[derive(Debug, Serialize, Deserialize)]
pub struct Variant {
    pub id: String,
    pub variant_calls: Vec<VariantCall>
}

impl Variant {
    pub fn new(id: String) -> Self {
        Self {
            id: id,
            variant_calls: Vec::new()
        }
    }

    pub fn add_variant_call(&mut self, variant_call: VariantCall) {
        // Check if chromosome_1 and chromosome_2 match existing VariantCall objects
        if self.variant_calls.len() > 0 {
            if self.variant_calls[0].chromosome_1 != variant_call.chromosome_1 {
                eprintln!("chromosome_1 must match chromosome_1 of an existing VariantCall object in this Variant object.");
                process::exit(exitcode::DATAERR);
            }
            if self.variant_calls[0].chromosome_2 != variant_call.chromosome_2 {
                eprintln!("chromosome_2 must match chromosome_2 of an existing VariantCall object in this Variant object.");
                process::exit(exitcode::DATAERR);
            }
        }

        // Find an index position to insert the new VariantCall object
        let insert_idx = self.variant_calls.binary_search_by(|item| item.position_1.cmp(&variant_call.position_1));
        match insert_idx {
            Ok(idx) => {
                self.variant_calls.insert(idx, variant_call);
            }
            Err(idx) => {
                self.variant_calls.insert(idx, variant_call);
            }
        }
    }

    pub fn get_mean(&self, attribute: String) -> f64 {
        let values: Vec<f64> = self.variant_calls.iter().map(|variant_call| {
            let value: f64 = match attribute.as_str() {
                "position_1" => variant_call.position_1 as f64,
                "position_2" => variant_call.position_2 as f64,
                "quality_score" => variant_call.quality_score as f64,
                "variant_size" => variant_call.variant_size as f64,
                "total_read_count" => variant_call.total_read_count as f64,
                "reference_allele_read_count" => variant_call.reference_allele_read_count as f64,
                "alternate_allele_read_count" => variant_call.alternate_allele_read_count as f64,
                "alternate_allele_fraction" => variant_call.alternate_allele_fraction as f64,
                _ => -1.0,
            };
            return value;
        }).collect();
        return utilities::calculate_mean(values);
    }

    pub fn get_median(&self, attribute: String) -> f64 {
        let values: Vec<f64> = self.variant_calls.iter().map(|variant_call| {
            let value: f64 = match attribute.as_str() {
                "position_1" => variant_call.position_1 as f64,
                "position_2" => variant_call.position_2 as f64,
                "quality_score" => variant_call.quality_score as f64,
                "variant_size" => variant_call.variant_size as f64,
                "total_read_count" => variant_call.total_read_count as f64,
                "reference_allele_read_count" => variant_call.reference_allele_read_count as f64,
                "alternate_allele_read_count" => variant_call.alternate_allele_read_count as f64,
                "alternate_allele_fraction" => variant_call.alternate_allele_fraction as f64,
                _ => -1.0,
            };
            return value;
        }).collect();
        return utilities::calculate_median(values);
    }

    pub fn get_max(&self, attribute: String) -> f64 {
        let values: Vec<f64> = self.variant_calls.iter().map(|variant_call| {
            let value: f64 = match attribute.as_str() {
                "position_1" => variant_call.position_1 as f64,
                "position_2" => variant_call.position_2 as f64,
                "quality_score" => variant_call.quality_score as f64,
                "variant_size" => variant_call.variant_size as f64,
                "total_read_count" => variant_call.total_read_count as f64,
                "reference_allele_read_count" => variant_call.reference_allele_read_count as f64,
                "alternate_allele_read_count" => variant_call.alternate_allele_read_count as f64,
                "alternate_allele_fraction" => variant_call.alternate_allele_fraction as f64,
                _ => -1.0,
            };
            return value;
        }).collect();
        return utilities::calculate_max(values);
    }

    pub fn get_min(&self, attribute: String) -> f64 {
        let values: Vec<f64> = self.variant_calls.iter().map(|variant_call| {
            let value: f64 = match attribute.as_str() {
                "position_1" => variant_call.position_1 as f64,
                "position_2" => variant_call.position_2 as f64,
                "quality_score" => variant_call.quality_score as f64,
                "variant_size" => variant_call.variant_size as f64,
                "total_read_count" => variant_call.total_read_count as f64,
                "reference_allele_read_count" => variant_call.reference_allele_read_count as f64,
                "alternate_allele_read_count" => variant_call.alternate_allele_read_count as f64,
                "alternate_allele_fraction" => variant_call.alternate_allele_fraction as f64,
                _ => -1.0,
            };
            return value;
        }).collect();
        return utilities::calculate_min(values);
    }
}

impl Clone for Variant {
    fn clone(&self) -> Self {
        let mut variant_calls: Vec<VariantCall> = Vec::new();
        for variant_call in &self.variant_calls {
            variant_calls.push(variant_call.clone());
        }
        Variant {
            id: self.id.clone(),
            variant_calls: variant_calls
        }
    }
}